modules/m_monitor.c

```diff
@@ -126,54 +100,61 @@ add_monitor(struct Client *client_p, const char *nicks)
 
 		rb_dlinkAddAlloc(client_p, &monptr->users);
 		rb_dlinkAddAlloc(monptr, &client_p->localClient->monitor_list);
+	}
 
-		if((target_p = find_named_person(name)) != NULL)
-		{
-			if(cur_onlen + strlen(target_p->name) +
-			   strlen(target_p->username) + strlen(target_p->host) + 3 >= BUFSIZE-3)
-			{
-				sendto_one(client_p, "%s", onbuf);
-				cur_onlen = mlen;
-				onptr = onbuf + mlen;
-			}
+	sendto_one_multiline_init(client_p, ",", form_str(RPL_MONONLINE),
+			me.name,
+			client_p->name,
+			"");
 
-			if(cur_onlen != mlen)
-			{
-				*onptr++ = ',';
-				cur_onlen++;
-			}
+	RB_DLINK_FOREACH_PREV(ptr, prev_head->prev)
+	{
+		monptr = ptr->data;
+		target_p = find_named_person(monptr->name);
 
-			arglen = sprintf(onptr, "%s!%s@%s",
-					target_p->name, target_p->username,
+		if (target_p != NULL)
+		{
+			sendto_one_multiline_item(client_p, "%s!%s@%s",
+					target_p->name,
+					target_p->username,
 					target_p->host);
-			onptr += arglen;
-			cur_onlen += arglen;
 		}
-		else
-		{
-			if(cur_offlen + strlen(name) + 1 >= BUFSIZE-3)
-			{
-				sendto_one(client_p, "%s", offbuf);
-				cur_offlen = mlen;
-				offptr = offbuf + mlen;
-			}
+	}
 
-			if(cur_offlen != mlen)
-			{
-				*offptr++ = ',';
-				cur_offlen++;
-			}
+	sendto_one_multiline_fini(client_p, NULL);
+
+	sendto_one_multiline_init(client_p, ",", form_str(RPL_MONOFFLINE),
+			me.name,
+			client_p->name,
+			"");
+
+	RB_DLINK_FOREACH_PREV(ptr, prev_head->prev)
+	{
+		monptr = ptr->data;
 
-			arglen = sprintf(offptr, "%s", name);
-			offptr += arglen;
-			cur_offlen += arglen;
+		if (find_named_person(monptr->name) == NULL)
+		{
+			sendto_one_multiline_item(client_p, "%s", monptr->name);
 		}
 	}
 
-	if(cur_onlen != mlen)
-		sendto_one(client_p, "%s", onbuf);
-	if(cur_offlen != mlen)
-		sendto_one(client_p, "%s", offbuf);
+	sendto_one_multiline_fini(client_p, NULL);
+
+	if (name)
+	{
+		char buf[DATALEN];
```

Comment:
I disagree with leaving this at 100, as it would significantly truncate the input if it's long (6 16 character nicks would be the point at which it would start truncating, which is why I increased it in the first place).  An estimated calculation of what the rest of the numeric would need for space leaves a bit more than 400 for this parameter (NICKLEN 32 * 2 + list limit 3 + numeric 3 + spaces and leading colon 6 + trailing string 21 = 97; 510 - 97 = 413), so that's what I'll use here